pkg/client/logs.go

```diff
@@ -111,30 +128,65 @@ func (r *Reader) Read(p []byte) (int, error) {
 	return len(data), nil
 }
 
-// getPodsForLogs retrieves the pods to stream logs from. If a plugin name has been provided, retrieve the pods with
+// getPodsToStreamLogs retrieves the pods to stream logs from. If a plugin name has been provided, retrieve the pods with
 // only the plugin label matching that plugin name. If no pods are found, or no plugin has been specified, retrieve
-// all pods within the namespace.
-func getPodsForLogs(client kubernetes.Interface, cfg *LogConfig) (*v1.PodList, error) {
+// all pods within the namespace. It will immediately return an error if unabel to list pods, but will otherwise
+// add pods onto the channel in a separate go routine so that this method does not block. It closes the pod channel once
+// all the pods have been reported.
+func getPodsToStreamLogs(client kubernetes.Interface, cfg *LogConfig, podCh chan *v1.Pod) error {
+	listOptions := metav1.ListOptions{}
 	if cfg.Plugin != "" {
 		selector := metav1.AddLabelToSelector(&metav1.LabelSelector{}, "sonobuoy-plugin", cfg.Plugin)
-		options := metav1.ListOptions{LabelSelector: metav1.FormatLabelSelector(selector)}
-		pods, err := client.CoreV1().Pods(cfg.Namespace).List(options)
-		if err != nil {
-			return nil, errors.Wrap(err, "failed to list pods")
-		}
+		listOptions = metav1.ListOptions{LabelSelector: metav1.FormatLabelSelector(selector)}
+	}
+
+	podList, err := client.CoreV1().Pods(cfg.Namespace).List(listOptions)
+	if err != nil {
+		return errors.Wrap(err, "failed to list pods")
+	}
 
-		if len(pods.Items) != 0 {
-			return pods, nil
+	go func() {
+		for _, p := range podList.Items {
+			podCh <- &p
 		}
+		close(podCh)
+	}()
+
+	return nil
+}
 
-		logrus.Warningf("failed to find pods for plugin %q, defaulting to all pods", cfg.Plugin)
```

Comment:
The diff here looks really strange; I removed this "defaulting to all pods" logic since it really made little since to me why we would want to do that if we were explicitly calling a method to look for a particular pod.  Let me know if you disagree; I did have to remove a testcase or 2 as a result of this but it didn't seem like intuitive behavior.

Reply:
Hm; looking in the blame/history I see you added this and it wasn't to really fallback in general but it was for a backwards compatibility reason. Let me know what your thoughts are on this now; I can and add it back if needed. (PR that added this here: https://github.com/vmware-tanzu/sonobuoy/pull/868)

Reply:
And it looks like that PR went in around the time of Sonobuoy 0.15.3 so it has been numerous versions since then. I just want to be sure we understand the risk/incompatibility we're actually facing.

Reply:
I decided to fall back to the default behaviour because that seemed like the safest thing to do, but it's still not what the user wants if they specified a plugin argument. My guess is that the warning about falling back to show all pods gets lost in the stream anyway, so perhaps it's just better to give an explicit error and bail out.